### PR TITLE
docs: SQLair tutorial updates

### PR DESCRIPTION
**Pull request by Aflynn50**
Update the tutorial to make it more procedural and possible to follow along.

### CODE REVIEW

File: docs/tutorial/new-tutorial.md [hunk 1]
```diff
@@ -0,0 +1,384 @@
+(tutorial)=
+# SQLair tutorial
+
+Imagine you are writing a Go program that handles your customers and their
+orders. The business logic is contained in the Go program but the customer and
+order data is stored in a SQL database. In this tutorial, we'll see how easy
+SQLair makes it to map the database between the database and the Go program.
+Let's get started!
```
Comment (tmihoc) [inline docs/tutorial/new-tutorial.md]: > map the database between the database and the

Sounds wrong or confusing.

File: docs/tutorial/new-tutorial.md [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+
+
+```{note}
+This tutorial will take around 30 minutes to complete. If you want a quicker
+start see {ref}`prepare-and-run`.
+```
```
Comment (tmihoc) [inline docs/tutorial/new-tutorial.md]: If that doc really acts as a quicker start, something is wrong there. In short, I wouldn't link there and also make sure that that doc cannot act as a quickstart.

File: docs/tutorial/new-tutorial.md [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+
+## Setting up the environment
```
Comment (tmihoc) [inline docs/tutorial/new-tutorial.md:16]: I'd rephrase as "Set up your test environment"

File: docs/tutorial/new-tutorial.md [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+
+To start with create a new folder called tutorial and initialise a new Go project.
```
Comment (tmihoc) [inline docs/tutorial/new-tutorial.md:18]:
```suggestion
To start with, create a new folder called `tutorial` and initialise a new Go project.
```

Comment (tmihoc) [reply docs/tutorial/new-tutorial.md:18]: In fact, I'd simplify even further:

```suggestion
The first thing you'll need is a Go project. Let's create one:
```

Then I'd add line by line comments in the code block, just to help someone who's really clueless catch up.

File: docs/tutorial/new-tutorial.md [hunk 5]
```diff
@@ -0,0 +1,384 @@
+(tutorial)=
+# SQLair tutorial
```
Comment (tmihoc) [inline docs/tutorial/new-tutorial.md]:
```suggestion
# Get started with SQLair
```

File: docs/tutorial/new-tutorial.md (hunk 3 above)
Comment (tmihoc) [inline docs/tutorial/new-tutorial.md]:
```suggestion
## Set up your test environment
```

File: docs/tutorial/new-tutorial.md [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+```
+
+## Setting up the environment
+
+To start with create a new folder called tutorial and initialise a new Go project.
+```bash
+mkdir tutorial 
+cd tutorial
+go mod init sqlair-tutorial
+```
```
Comment (tmihoc) [inline docs/tutorial/new-tutorial.md]: Not sure how to create a suggestion for multiple lines, but here's what I'd do:

```suggestion
The first thing you'll need is a Go project. Let's create one:
```bash
# Create a new directory:
mkdir tutorial 

# Enter the directory:
cd tutorial

# Initialise a new Go project:
go mod init sqlair-tutorial
```
```

PS Based on the intro of the tutorial, it sounds like what the user really needs is a Go program and a database. So I'm not sure why we're just getting them to initialise the project here.

File: docs/tutorial/new-tutorial.md [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+
+We first need to set up the backbone of our project, the database. The easiest
```
Comment (tmihoc) [inline docs/tutorial/new-tutorial.md]: This tutorial "we" sounds like a good idea but it's usually either awkward (almost always) or confusing (e.g., when you need to distinguish between "here's the output I got -- yours might be similar"). I'd mostly stick to "you".

File: docs/tutorial/new-tutorial.md [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+way to create a SQL database in Go is to use [SQLite](https://www.sqlite.org/)
+with the [Go driver from mattn](https://github.com/mattn/go-sqlite3) though
+SQLair works with any database driver that can interface with
+[database/sql](https://pkg.go.dev/database/sql). With SQLite, we can create an in
+process, in memory database that can be populated within the program and that
+will disappear when the program finishes running.
+
+We also want to set the database up with our customer and order schema. We will
+have two tables. One for customers and the other for orders.
+
```
Comment (tmihoc) [inline docs/tutorial/new-tutorial.md]: This whole paragraph is too much talking at the user and it's also too much detail about something that is not at issue. I'd simplify ("You'll also need a SQL database to store your customer and order data. Let's go ahead and create one:", then in-code comments to clarify the steps, e.g., that there are many ways to create a SQL database but, since our program is in Go, we'll want Go tooling). If you still feel the need to highlight something else (e.g., that SQLair works with any database driver that can interface with... -- though I'd also wonder why we're not highlighting that bit in the homepage), put that in a note (which, if you want, you can make exciting by prefixing it with "Did you know? ...").

File: docs/tutorial/new-tutorial.md [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+To create the database, create a new `main.go` file in the project, open it in
+your favourite editor and copy in the code below.
```
Comment (tmihoc) [inline docs/tutorial/new-tutorial.md:37]: Again, not sure how to do multi-line suggestions, but here what's I'd do: 

```suggestion
To create the database, in your Go project create a `main.go` file; open it in your favourite editor; then copy in the code below:
```

File: docs/tutorial/new-tutorial.md [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+    err := tutorial()
+    if err != nil {
+        fmt.Printf("ERROR: %s\n", err)
+    }
+    fmt.Println("Finished without errors!")
+}
+```
+The program above will open the SQLite database, add our schema to it, then wrap
+it as a SQLair DB. SQLair does not create the underlying database, only provide
+an interface for getting and setting the values within it.
```
Comment (tmihoc) [inline docs/tutorial/new-tutorial.md]: I'd do this whole thing as below (changed information order, updated punctuation, added comment on the outcome before the code, demoted by-the-way info to a note, and fixed some typos):

```{suggestion}
To create the database, in your Go project create a `main.go` file; open it in your favourite editor; then copy in the code below. 

<code block>

This will open the SQLite database, add your schema to it, then wrap it as a SQLair DB. 

```{note}
 SQLair does not create the underlying database -- it only provides an interface for getting and setting the values within it. 
```
```

File: docs/tutorial/new-tutorial.md [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+Now, to test the program, install the dependencies and run:
+```bash
+go mod tidy
+go run .
+```
+You should get the output: `"Finished without errors!"`.
+
+# Defining the data types
+
+Before we start processing our customer and order data we need to define go
```
Comment (tmihoc) [inline docs/tutorial/new-tutorial.md]: Try to avoid this sort of "Before we start..." sentences. The user doesn't gain much for them and you're just wasting their attention on meta-talk.

File: docs/tutorial/new-tutorial.md (hunk 2 above)
Comment (Aflynn50) [reply docs/tutorial/new-tutorial.md]: That doc just shows you how to prepare and run a query, which may be all they want to do. I guess its not really a quick start, just a proof of concept kind of thing. I'll remove the link.

File: docs/tutorial/new-tutorial.md [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+structs that we are going to use to represent them in our program. These should
+map onto the rows of the tables in the database.
+
+Add the following types to the `main.go` file in your tutorial project.
```
Comment (tmihoc) [inline docs/tutorial/new-tutorial.md]: Try to put the "where things must happen" before the "what must happen". So:

> In your project's `main.go` file add the types below:

File: docs/tutorial/new-tutorial.md (hunk 7 above)
Comment (Aflynn50) [reply docs/tutorial/new-tutorial.md]: Cool, will change that.

File: docs/tutorial/new-tutorial.md (hunk 8 above)
Comment (Aflynn50) [reply docs/tutorial/new-tutorial.md]: I've compressed that first paragraph to just mention the database and driver we are using. I think this is pretty important backgroud info since it affects the syntax of the SQL that people would write. I've moved the comment about in memory/in process into the code.

The point about it working with any driver should maybe be on the homepage, but I guess its kindof implicit.

File: docs/tutorial/new-tutorial.md (hunk 11 above)
Comment (Aflynn50) [reply docs/tutorial/new-tutorial.md]: Makes sense, I can pretty much halve the size of this paragraph and still say the same thing.

**Review by tmihoc — CHANGES_REQUESTED** [review]
As discussed during our meeting:

(1) The tutorial should have the following main sections:

* Set things up
This should say something like "You'll need Go, a SQL database created with Go, Go structs tagged for the database [not sure about this one but it feels like a natural way to put it], and SQLair. 
* Write data to your database
* Read data from your database
The read all bit can be done in the sense of a homework challenge, with a link to the relevant reference and an expand block with what you have now as the answer.
* Tear things down
This should include deleting the project, the library, uninstalling Go (even if you only mention these things verbally or link to external stuff).

Note: I wouldn't bother with any subsections. 

(2) Something the structure suggested just now already addresses: I'd keep each section active. (That will also help avoid the temptation to slide into how-to guide or reference mode.) The sections should follow a pattern with a little bit of main text which should also highlight what needs to be done and what that will achieve + code with comments that help people understand the details. 

(3) If it needs to link to anything, the tutorial should only link in the sense of "See more" (an out-of-text, end-of-chapter "Further reading" sort of thing). Whenever it links to a different website, the link should be prefixed with that (e.g., [Go | Struct tags]...). 

I think if we make these changes the tutorial will be really strong.

**Review by tmihoc — CHANGES_REQUESTED** [review]
Summary of our in-person chat:

(1) Make the titles Verb rather than Verb-ing based.

(2) Change the setup and teardown titles to Set things up, Tear things down.

(3) In the setup section, keep the pieces where you use sql and sqlDB (creating the SQL database) but move the part where you wrap that DB into a SQLair db to the next section where you prepare the Go types -- those two topics are in the same umbrella (prep work, but not the setup kind but rather the kind that helps you get into the SQLair logic.

(4) In the setup bit, be more explicit about the bits that matter and cut out anything that does not: "You'll need Go version ..., a Go program that does... and its associated SQL database that stores... First, ensure you have the right version of Go: .. . Next, create a directory ... and in it create a file with your Go program that also creates the SQL database: ... You're all set -- time to get cracking!

I'll keep thinking about the titles for the sections between Set things up and Tear things down. (It's a bit like `juju bootstrap` -- in the tutorial these days we don't highlight it too much; rather, what we emphasize is "this is what your deploy phase looks like".)

File: docs/tutorial/tutorial.md [hunk 13]
```diff
@@ -1,305 +1,495 @@
...
+> See more: [`Query.GetAll`](https://pkg.go.dev/github.com/canonical/sqlair#Query.GetAll)
 
-If you have any more in depth questions or issues please use see our [GitHub](https://github.com/canonical/sqlair). Contributions are always welcome! Hopefully this has been of some use, enjoy using the library.
+## Cleanup
+Getting rid of the tutorial is as simple as deleting the directory! No further
+steps required.
+```bash
+cd ..
+rm -r tutorial
+```
```
Comment (tmihoc) [inline docs/tutorial/tutorial.md:490]: What about uninstalling the SQLair library?

Comment (Aflynn50) [reply docs/tutorial/tutorial.md:490]: Thats a good point, and as far as I can tell its surprisingly difficult. There's a stack overflow post [here](https://stackoverflow.com/questions/13792254/removing-packages-installed-with-go-get) about it, but the top answer (using `go get package@none`) no longer works because in the latest version they made `go get` only work if you're already in a project.

It seems you can do this:
```
go clean -cache -modcache
```
Which I think removes unused packages but when I ran it on my machine it came back with a weird message about linking. I'm not sure I'd want to put it in the tutorial.

In general with Go I think people just don't really worry about this sort of thing, but I agree it would be nice to have a way to do this.

**Review by tmihoc — APPROVED** [review]
Look as discussed, thanks!
